src/core/core.datasetController.js

```diff
@@ -1189,3 +1190,23 @@ DatasetController.prototype.dataElementOptions = [
 	'borderWidth',
 	'pointStyle'
 ];
+
+/**
+ * @type {{[key: string]: typeof DatasetController}}
+ */
+export const controllers = {};
+
+export function registerController(controller) {
+	if (typeof controllers[controller.id] !== 'undefined') {
+		// already registered
```

Comment:
throw that?

Reply:
I wouldn't e.g., a PieController will auto register a DougnutController since it is its parent, but might be already registered.